imbi/endpoints/integrations/github.py

```diff
@@ -0,0 +1,139 @@
+            'grant_type': 'authorization_code',
+            'redirect_uri': str(self.integration.callback_url),
+            'code': code,
+        }
+        response = await self.http_fetch(
+            str(self.integration.token_endpoint),
+            method='POST',
+            body=body,
+            content_type=sprockets.mixins.http.CONTENT_TYPE_JSON,
+            request_headers={
+                'X-GitHub-Api-Version': '2022-11-28',
+                'Accept': sprockets.mixins.http.CONTENT_TYPE_JSON
+            })
+        if not response.ok:
+            self.logger.error('failed to exchange auth code for token: %s %s',
+                              response.body['error'],
+                              response.body['error_description'])
+            raise errors.InternalServerError(
+                'failed exchange auth code for token: %s',
+                response.code,
+                title='GitHub authorization failure',
+                instance={
+                    'error': response.body['error'],
+                    'error_description': response.body['error_description'],
+                })
+        self.logger.debug('response_body %r', response.body)
+        return GitHubToken(access_token=response.body['access_token'],
+                           refresh_token=response.body['refresh_token'])
+
+    async def fetch_github_user(self, token: GitHubToken) \
+            -> typing.Tuple[int, str]:
+        response = await self.http_fetch(
+            str(self.integration.api_endpoint / 'user'),
+            request_headers={
+                'X-GitHub-Api-Version': '2022-11-28',
+                'Accept': sprockets.mixins.http.CONTENT_TYPE_JSON,
+                'Authorization': f'Bearer {token.access_token}'
+            })
+        if response.ok:
+            return (response.body['id'], response.body['email'])
+        raise errors.InternalServerError(
+            'failed to retrieve GitHub user from access token',
+            title='GitHub user lookup failure')
+
+    async def revoke_github_token(self, token: GitHubToken, username: str):
+        credentials = []
+        token_types = []
+        if token.access_token:
+            credentials.append(token.access_token)
+            token_types.append('access_token')
+        if token.refresh_token:
+            credentials.append(token.refresh_token)
+            token_types.append('refresh_token')
+
+        if not credentials:
+            return
+
+        body = {'credentials': credentials}
+        response = await self.http_fetch(
+            str(self.integration.api_endpoint / 'credentials' / 'revoke'),
+            method='POST',
+            body=body,
+            content_type=sprockets.mixins.http.CONTENT_TYPE_JSON,
+            request_headers={
+                'X-GitHub-Api-Version': '2022-11-28',
+                'Accept': sprockets.mixins.http.CONTENT_TYPE_JSON,
+            },
+        )
```

Comment:
I would have expected the OAuth 2 client ID and secret to show up here somewhere. I haven't looked at the API yet.

Reply:
You pass the token(s) in the `credentials` field of the body: https://docs.github.com/en/rest/credentials/revoke?apiVersion=2022-11-28#revoke-a-list-of-credentials

I agree I was a bit surprised to see that too